### DISCUSSION TITLE
custom backend

### DISCUSSION THREAD

**Discussion by djsteffey** [x64dbg/x64dbg#3458] [Q&A] [answered]
I am interested in making a custom backend debugger that uses the x64dbg frontend.  Think of it as being able to support binaries other than like your standard windows exe, maybe even an emulation environment.

I have looked through some of the source code for x64dbg and it looks like this concept may be supported under the idea of a "debug engine"?  I see there is support right now for a Glee and a TitanEngine.

How it looks like x64dbg is structured is that the front end will makes calls to this backend engine to actual pull the values and information it wants to display.  Or send to this backend the values that it wants to modify.  Is this correct?

If so, is there any documentation on making a custom engine (ie what all functions and parameters need to be exported) and on how to load this in x64dbg.

**Comment by ZehMatt** [comment, marked as answer]
You can check out GleeBug which is more or less what you describe, but its also not pretty at the same time as it was never designed to be like this.

**Reply by djsteffey** [reply]
Thanks.  I will check it out.  I do see that the choices for the debug engine are hard coded values in x64dbg itself.  Easy enough for me to modify that add add a selection for what I want to build.  But it would mean if others (probably not lol) would want to use my debug engine they would also have to modify and compile x64dbg to accommodate another selection.

I dont suppose x64dbg repo here could put that on the todo list?  to support generic debug engine like they do plugins.

Maybe I will open an issue for that.

**Reply by mrexodia** [reply]
I think the easiest workaround will be to just use hooking in the plugin 🤷🏻‍♂️. It's a very unlikely and niche use case, so hardly worth the maintenance effort...
